api/alerts.go

```diff
@@ -0,0 +1,207 @@
+
+	body, readErr := ioutil.ReadAll(res.Body)
+	if readErr != nil {
+		log.Fatal(readErr)
+	}
+
+	alerts := []Alert{}
+	jsonErr := json.Unmarshal(body, &alerts)
+	if jsonErr != nil {
+		log.Fatal(jsonErr)
+	}
+
+	return alerts, nil
+}
+
+func (a *Alerts) Update(viewName string, alert *Alert) (*Alert, error) {
+	existingID, err := a.convertAlertNameToID(viewName, alert.Name)
+	if err != nil {
+		return nil, fmt.Errorf("could not convert alert name to id: %v", err)
+	}
+
+	url := fmt.Sprintf("api/v1/repositories/%s/alerts/%s", viewName, existingID)
+
+	jsonStr, err := json.Marshal(alert)
+	if err != nil {
+		return nil, fmt.Errorf("unable to convert alert to json string: %v", err)
+	}
+	// Humio requires notifiers to be specified even if no notifier is desired
+	if alert.Notifiers == nil {
+		alert.Notifiers = []string{}
+	}
+	res, postErr := a.client.HttpPUT(url, bytes.NewBuffer(jsonStr))
+	if postErr != nil {
+		return nil, fmt.Errorf("could not add alert in view %s with name %s, got: %v", viewName, alert.Name, postErr)
+	}
+
+	body, readErr := ioutil.ReadAll(res.Body)
+	if readErr != nil {
+		log.Fatal(readErr)
+	}
+
+	resAlert := Alert{}
+	jsonErr := json.Unmarshal(body, &resAlert)
+	if jsonErr != nil {
+		log.Fatal(jsonErr)
+	}
+	return &resAlert, nil
+}
+
+func (a *Alerts) Add(viewName string, alert *Alert, updateExisting bool) (*Alert, error) {
+	nameAlreadyInUse, err := a.alertNameInUse(viewName, alert.Name)
+	if err != nil {
+		return nil, fmt.Errorf("could not determine if alert name is in use: %v", err)
+	}
+	if nameAlreadyInUse {
+		if updateExisting == false {
+			return nil, fmt.Errorf("alert with name %s already exists", alert.Name)
+		}
+		return a.Update(viewName, alert)
+	}
+
+	url := fmt.Sprintf("api/v1/repositories/%s/alerts/", viewName)
+	jsonStr, err := json.Marshal(alert)
+	if err != nil {
+		return nil, fmt.Errorf("unable to convert alert to json string: %v", err)
+	}
+	// Humio requires notifiers to be specified even if no notifier is desired
+	if alert.Notifiers == nil {
+		alert.Notifiers = []string{}
+	}
+	res, postErr := a.client.HttpPOST(url, bytes.NewBuffer(jsonStr))
+	if postErr != nil {
+		return nil, fmt.Errorf("could not add alert in view %s with name %s, got: %v", viewName, alert.Name, postErr)
+	}
+
+	body, readErr := ioutil.ReadAll(res.Body)
+	if readErr != nil {
+		log.Fatal(readErr)
+	}
+
+	resAlert := Alert{}
+	jsonErr := json.Unmarshal(body, &resAlert)
+	if jsonErr != nil {
+		log.Fatal(jsonErr)
+	}
+	return &resAlert, nil
+}
+
+func (a *Alerts) Get(view, name string) (*Alert, error) {
+	alertID, err := a.convertAlertNameToID(view, name)
+	if err != nil {
+		return nil, fmt.Errorf("could not find a notifier in view %s with name: %s", view, name)
+	}
+
+	url := fmt.Sprintf("api/v1/repositories/%s/alerts/%s", view, alertID)
+
+	res, err := a.client.HttpGET(url)
+	if err != nil {
+		return nil, fmt.Errorf("could not get alert with id %s, got: %v", alertID, err)
+	}
+
+	body, readErr := ioutil.ReadAll(res.Body)
```

Comment:
Lines 145-154 could also be broken down to a private method.

Reply:
I've pushed a commit that pulls that code out to a private method.